[Home] List of Characters per Category for Home Screen

Fernando-Guerrero-jala opened the pull request:
- Add categories sections in a table view that includes title label, see all button and horizontal scrollable collection view
- Load data from API into ViewModel
- Show character portraits in collection view
- Update .gitignore
- Use segue transition to details screen for each character portrait
- Add colors helper
- Match UI with Figma protoype

Fernando-Guerrero-jala committed:
- Add one category section that includes title label, see all button and horizontal scrollable collection view
- Load data from API into ViewModel
- Show character portraits in collection view

Fernando-Guerrero-jala committed:
- Add table view to show each category section
- Add images to Assets

Fernando-Guerrero-jala committed:
Merge branch 'develop'

Conflicts:
	MarvelCharsApp/MarvelCharsApp/Views/HomeViewController.swift

Fernando-Guerrero-jala committed:
- Update API Delegate implementation in ViewModel
- Display character portraits per category

Fernando-Guerrero-jala committed:
- Add transition to character portraits
- Update code

Fernando-Guerrero-jala committed:
- Update fonts configuration for Xcode

Fernando-Guerrero-jala committed:
Update .gitignore

Fernando-Guerrero-jala committed:
Delete .DS_Store from repository accidentally added

Fernando-Guerrero-jala committed:
- Update category row implementation.

Fernando-Guerrero-jala committed:
Merge branch 'develop'

Conflicts:
	MarvelCharsApp/MarvelCharsApp.xcodeproj/project.pbxproj
	MarvelCharsApp/MarvelCharsApp/Views/Storyboards/Base.lproj/Main.storyboard

Fernando-Guerrero-jala committed:
- Add Colors Helper
- Update Models for new API JSON
- Update Views with new assets
- Update transition to Details with segue
- Add Unit tests for View Model and Models

Fernando-Guerrero-jala committed:
- Update UI to match Figma protoype

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Models/Entities/CategoryModel.swift:
```diff
@@ -20,7 +20,10 @@ struct CategoryModel {
-    func getCharacter(at index: Int) -> CharacterModel {
-        return self.characters[index]
+    func getCharacter(at index: Int) -> CharacterModel? {
+      if index<0 || index>=charactersCount || characters.isEmpty {
```
don't forget code indentation

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Models/Entities/MarvelAPIModel.swift:
```diff
@@ -51,7 +51,43 @@ enum Unity: String, Codable {
 typealias Category = [CategoryElement]
+
+// MARK: - Shorter inits
+extension Character {
```
Add the code inside the Character code, there's no need for an extension

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Models/Entities/MarvelAPIModel.swift:
```diff
@@ -51,7 +51,43 @@ enum Unity: String, Codable {
+    self.movies = movies
+  }
+}
+extension Caracteristics {
```
same with all this inits that has the structs on the same file

jmacboy commented on MarvelCharsApp/MarvelCharsApp/ViewModels/HomeViewModel.swift:
```diff
@@ -7,6 +7,42 @@
+  var categories: [CategoryModel] = []
+  var categoriesNames: [String] = []
+  
+  init(){
```
you're forgetting to indent your file

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Views/Cells/CategoryRowCell.swift:
```diff
@@ -0,0 +1,80 @@
+    super.awakeFromNib()
+    self.selectionStyle = .none
+    
+    charactersCollectionView.register(UINib(nibName: "CharacterPortraitCell", bundle: nil), forCellWithReuseIdentifier: "CharacterPortraitID")
```
this identifier should be stored on a constant

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Views/Cells/CategoryRowCell.swift:
```diff
@@ -0,0 +1,80 @@
+  }
+  
+  func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    let cell = charactersCollectionView.dequeueReusableCell(withReuseIdentifier: "CharacterPortraitID", for: indexPath) as! CharacterPortraitCell
```
also this identifier as a constant

jmacboy replied:
You should avoid using force casting, is a bad practice.
I'm referring to as!

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Views/HomeViewController.swift:
```diff
@@ -8,20 +8,69 @@
+  }
+  
+  func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+    let cell = self.tableView.dequeueReusableCell(withIdentifier: "CategoryRowID", for: indexPath) as! CategoryRowCell
```
always use the identifiers as constants

jmacboy commented on MarvelCharsApp/MarvelCharsApp/Views/HomeViewController.swift:
```diff
@@ -8,20 +8,69 @@
+extension HomeViewController: HomeViewModelDelegate {
+  func onFetchDataSuccessfully(categories: [CategoryModel]) {
+    refresh()
+    return
```
Do you need this return for something?

jmacboy commented on MarvelCharsApp/MarvelCharsAppTests/CategoryModelTest.swift:
```diff
@@ -0,0 +1,78 @@
+    XCTAssertNil(category.getCharacter(at: 0))
+  }
+  
+  func testGetCharacterThrowsErrorWhenIndexOutOfRange() throws {
```
The name says throw error, but I see that you're validating nils. Change the name to reflect that or make the function throw an exception as the name says.

jmacboy commented on MarvelCharsApp/MarvelCharsAppTests/CategoryModelTest.swift:
```diff
@@ -0,0 +1,78 @@
+    XCTAssertEqual(character?.alterEgo, "Tony Stark")
+  }
+  
+  func testGetCharacterThrowsErrorWhenNotSet() throws {
```
The name says throw error, but I see that you're validating nils. Change the name to reflect that or make the function throw an exception as the name says.

jmacboy commented on MarvelCharsApp/MarvelCharsAppTests/HomeViewModelTest.swift:
```diff
@@ -0,0 +1,44 @@
+  
+  func testReturnNilWhenGetCategoryWithWrongName() throws {
+    let viewModel = HomeViewModel()
+    viewModel.categories.append(CategoryModel(category: "Heroes"))
```
The data should not be mocked this way on the viewModel. You need to create a mock Manager and pass it as a value to the model. The viewmodel should call fetchData here and call this manager to obtain the fake data.
The declaration of that fake manager can be in this class.

jmacboy commented on MarvelCharsApp/MarvelCharsAppTests/HomeViewModelTest.swift:
```diff
@@ -0,0 +1,44 @@
+
+  func testGetCategoryWithName() throws {
+    let viewModel = HomeViewModel()
+    viewModel.categories.append(CategoryModel(category: "Heroes"))
```
same as the other method

jmacboy commented on MarvelCharsApp/MarvelCharsAppTests/HomeViewModelTest.swift:
```diff
@@ -0,0 +1,44 @@
+  
+  func testReturnNilWhenGetCategoryWithNameAndCategoriesIsEmpty() throws {
+    let viewModel = HomeViewModel()
+    let category = viewModel.getCategoryWith(name: "")
```
same as the other method, you need to call Fetchdata

Fernando-Guerrero-jala committed:
- Fix indentation
- Add extension inits to structs
- Update view model to set manager for testing
- Update tests names
- Add mocked manager for testing
- Add identifiers properties to classes
- Fix force casting

Fernando-Guerrero-jala committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	MarvelCharsApp/MarvelCharsApp.xcodeproj/project.pbxproj

Fernando-Guerrero-jala committed:
Merge branch 'develop' into feature/MAR-4-CharactersLists

Fernando-Guerrero-jala committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/MAR-4-CharactersLists'

Fernando-Guerrero-jala committed:
commit

Fernando-Guerrero-jala committed:
Merge branch 'develop'

Conflicts:
	MarvelCharsApp/MarvelCharsApp.xcodeproj/project.pbxproj

Fernando-Guerrero-jala committed:
- Fix Storyboard

Fernando-Guerrero-jala committed:
Merge remote-tracking branch 'refs/remotes/origin/develop'

Conflicts:
	.gitignore

Fernando-Guerrero-jala committed:
Delete UIImageHelper

Fernando-Guerrero-jala committed:
Merge with develop

Fernando-Guerrero-jala committed:
- Fix project file
- Fix storyboard merge
